[WIP] Simplify core

LeaVerou opened the pull request.

LeaVerou committed:
[WIP] Simplify core

LeaVerou committed:
highlightAll

LeaVerou committed:
Just use properties on env, not a separate `state` property

LeaVerou committed:
Move singleton, eliminate state further

LeaVerou committed:
Finish removing HookState

LeaVerou committed:
Update prism-class.ts

LeaVerou committed:
Delete hook-state.ts

LeaVerou committed:
Move more functions out of prism-class

LeaVerou committed:
Merge branch 'v2' into simplify

LeaVerou committed:
Move tokenize

DmitrySharabin commented on src/core/hooks.ts:
```diff
@@ -60,102 +56,4 @@ export class Hooks {
-	attributes: Record<string, string>;
-	language: string;
-}
+export type HookCallback<HookName extends string> = (env: Record<string, any>) => void;
```
Do we need `HookName` here? It doesn't seem to be used anywhere? Does callback depend on the hook name it's associated with?

LeaVerou replied:
No, it's unused.

DmitrySharabin commented on src/core/hooks.ts:
```diff
@@ -1,7 +1,3 @@
-
 export class Hooks {
 	// eslint-disable-next-line func-call-spacing
 	private _all = new Map<string, ((env: unknown) => void)[]>();
```
What if we switch to native private properties and methods? I think TS will still correctly infer the property visibility. Or is there a reason to stick with the `_` convention?

LeaVerou replied:
`_all` indicates that it's private, but is still readable by any code doing some weird low-level thing. That said, it doesn't seem to be read by any other part of the codebase, so we could, though it also doesn't provide much of a benefit so I don't think it's particularly important, so I'm inclined to leave it as-is.

DmitrySharabin commented:
I can't comment on the corresponding line since it's not in the PR, but I can't understand why we need a `list` variable if we already have `hooks`. Moreover, we do `const list = hooks`. What am I missing?

```ts
add<Name extends string> (name: Name, callback: HookCallback<Name>): () => void {
	let hooks = this._all.get(name);
	if (hooks === undefined) {
		hooks = [];
		this._all.set(name, hooks);
	}
	const list = hooks;

	list.push(callback as never);

	return () => {
		const index = list.indexOf(callback as never);
		if (index !== -1) {
			list.splice(index, 1);
		}
	};
}
```

DmitrySharabin commented on src/core/registry.ts:
```diff
@@ -1,7 +1,7 @@
 import type { ComponentProto, Grammar } from '../types';
-import type { Prism } from './prism';
+import type { Prism } from './prism-class';
 
```
```suggestion
import type Prism from './prism-class';
```

Another option would be:

```ts
import type { Prism } from './prism';
```

LeaVerou replied:
We want to import from the lowest level module we can

DmitrySharabin commented on src/core.ts:
```diff
@@ -1,2 +1,2 @@
-export { Prism } from './core/prism';
+export { Prism } from './core/prism-class';
```
```suggestion
export { default as Prism } from './core/prism-class';
```

DmitrySharabin commented on src/global.ts:
```diff
@@ -1,10 +1,8 @@
-import { Prism } from './core/prism';
+import globalPrism, { Prism } from './core/prism-class';
```
```suggestion
import globalPrism, { type Prism } from './core/prism';
```

DmitrySharabin commented on src/global.ts:
```diff
@@ -1,10 +1,8 @@
-const globalPrism = (namespace[globalSymbol] ??= new Prism());
+declare global {
+	var Prism: Prism | undefined
+}
```
```suggestion
	const Prism: Prism | undefined;
```

Is there a reason for using `var` I might miss?

LeaVerou replied:
Nope, that's just what the code example I found for extending `global` was using 😛 But this suggestion is also on the wrong line.

DmitrySharabin commented on src/types.d.ts:
```diff
@@ -1,4 +1,4 @@
-import type { Prism } from './core/prism';
+import type { Prism } from './core/prism-class';
```
```suggestion
import type Prism from './core/prism-class';
```

DmitrySharabin commented on src/core/prism-class.ts:
```diff
@@ -0,0 +1,49 @@
+ */
+export default class Prism {
+	hooks = new Hooks();
+	components = new Registry(this);
```
It looks like we have only languages there (plugins have their own property). Why not use languages instead of components here?

LeaVerou replied:
Agreed. Can you do it?

DmitrySharabin replied:
Sure, will do

DmitrySharabin commented on src/core/prism-class.ts:
```diff
@@ -0,0 +1,49 @@
+	 * See {@link highlight}
+	 */
+	highlight (text: string, language: string, options?: HighlightOptions): string {
+		return highlight.call(this, text, language, options);
```
```suggestion
	highlight (text: string, language: string, options: HighlightOptions = {}): string {
```

LeaVerou replied:
I think this suggestion is on the wrong line

DmitrySharabin replied:
It's. Weird ... I'll fix it

LeaVerou committed:
Update src/core.ts

Co-authored-by: Dmitry Sharabin <dmitrysharabin@gmail.com>

LeaVerou committed:
Update src/types.d.ts

Co-authored-by: Dmitry Sharabin <dmitrysharabin@gmail.com>

DmitrySharabin commented on src/core/prism-class.ts:
```diff
@@ -0,0 +1,49 @@
+	/**
+	 * See {@link highlight}
+	 */
+	highlight (text: string, language: string, options?: HighlightOptions): string {
```
```suggestion
	highlight (text: string, language: string, options: HighlightOptions = {}): string {
```

LeaVerou committed:
Update src/global.ts

Co-authored-by: Dmitry Sharabin <dmitrysharabin@gmail.com>

LeaVerou committed:
Update hooks.ts

DmitrySharabin commented on src/global.ts:
```diff
@@ -1,10 +1,8 @@
-const namespace = globalThis as Partial<Record<typeof globalSymbol, Prism>>;
-const globalPrism = (namespace[globalSymbol] ??= new Prism());
+declare global {
+	var Prism: Prism | undefined
```
```suggestion
	const Prism: Prism | undefined;
```

LeaVerou committed:
Update hooks.ts

LeaVerou committed:
Update src/core/prism-class.ts

Co-authored-by: Dmitry Sharabin <dmitrysharabin@gmail.com>

LeaVerou committed:
Update src/global.ts

Co-authored-by: Dmitry Sharabin <dmitrysharabin@gmail.com>

LeaVerou committed:
Update registry.ts

DmitrySharabin commented on src/core/stringify.ts:
```diff
@@ -0,0 +1,58 @@
+	}
+
+	const env: Record<string, any> & { classes: string[]; attributes: Record<string, string> } = {
+		type: o.type,
```
`o` is a string. Does it have a `type` property?

LeaVerou replied:
It's not just a string. If it's a string, we've already returned in line 18.

DmitrySharabin replied:
Oh, right. Here, it's a `TokenStream` or a `Token`. My bad.

DmitrySharabin commented on src/core/stringify.ts:
```diff
@@ -0,0 +1,58 @@
+		type: o.type,
+		content: stringify(o.content, language, hooks),
+		tag: 'span',
+		classes: ['token', o.type],
```
And here

DmitrySharabin commented on src/core/stringify.ts:
```diff
@@ -0,0 +1,58 @@
+		language,
+	};
+
+	const aliases = o.alias;
```
And here

DmitrySharabin commented on src/core/highlight.ts:
```diff
@@ -0,0 +1,55 @@
+	this: Prism,
+	text: string,
+	language: string,
+	options?: HighlightOptions
```
```suggestion
	options: HighlightOptions = {}
```

DmitrySharabin commented on src/core/highlight.ts:
```diff
@@ -0,0 +1,55 @@
+	const prism = this ?? singleton;
+
+	const languageId = prism.components.resolveAlias(language);
+	const grammar = options?.grammar ?? prism.components.getLanguage(languageId);
```
```suggestion
	const grammar = options.grammar ?? prism.components.getLanguage(languageId);
```

We don't need this if we accept the previous suggestion.

LeaVerou committed:
Update eslint.config.mjs

LeaVerou committed:
Remove known plugins

LeaVerou committed:
Update prism.ts

LeaVerou committed:
Update src/core/highlight.ts

Co-authored-by: Dmitry Sharabin <dmitrysharabin@gmail.com>

LeaVerou committed:
Update src/core/highlight.ts

Co-authored-by: Dmitry Sharabin <dmitrysharabin@gmail.com>

DmitrySharabin commented on src/core/stringify.ts:
```diff
@@ -0,0 +1,58 @@
+	return `<${env.tag} class="${env.classes.join(' ')}"${attributes}>${env.content}</${env.tag}>`;
+}
+
+export { stringify };
+export default stringify;
```
The answer to my following question will probably become TIL. Why do we need to export the same thing in two different ways? To make different ways of importing possible?

LeaVerou replied:
Yes

DmitrySharabin commented on src/plugins/keep-markup/prism-keep-markup.ts:
```diff
@@ -77,11 +76,11 @@ export default {
 			},
 			'after-highlight': (env) => {
-				const data = env.state.get(markupData, []);
+				const data = env.markupdata ?? [];
```
```suggestion
				const data = env.markupData ?? [];
```

LeaVerou committed:
Merge branch 'v2' into simplify

LeaVerou committed:
Merge branch 'simplify' of https://github.com/PrismJS/prism into simplify

LeaVerou committed:
Update src/plugins/keep-markup/prism-keep-markup.ts

Co-authored-by: Dmitry Sharabin <dmitrysharabin@gmail.com>

DmitrySharabin commented on src/known-plugins.d.ts:
We also need to delete a mention of this file from `scripts/build.ts` (line 332)

LeaVerou replied:
Can you do it?

DmitrySharabin replied:
Done

DmitrySharabin committed:
[build] Delete `known-plugins.d.ts`

LeaVerou committed:
[WIP] Make language definitions more declarative, simplify registry, better modularity

LeaVerou committed:
Merge branch 'simplify' of https://github.com/PrismJS/prism into simplify

DmitrySharabin commented on eslint.config.mjs:
```diff
@@ -171,6 +171,10 @@ const config = [
+			'@typescript-eslint/no-unsafe-assignment': 'off',
+			'@typescript-eslint/no-unsafe-member-access': 'off',
+			'@typescript-eslint/no-unsafe-return': 'off',
+			'@typescript-eslint/no-unsafe-argument': 'off',
```
Can't comment on line 182 (it's not in the PR), but I'd suggest adding `src/config.ts` to the list of (core) files to calm ESLint down about the undefined `document`.

LeaVerou replied:
I wonder if we need to tweak the config somehow, I would expect `requestAnimationFrame()` would be treated as a known global.

DmitrySharabin commented on src/config.ts:
```diff
@@ -0,0 +1,65 @@
+const scriptElement: HTMLScriptElement = hasDOM && document.currentScript;
+// @ts-ignore
+const globalConfig: Record<String, string | string[] | null> =
+	globalThis.Prism?.constructor?.name === 'Object' ? globalThis.Prism : {};
```
```suggestion
const globalConfig: Record<string, string | string[] | null> =
```

Ignore my suggestion if using uppercase `String` was intentional.

DmitrySharabin replied:
```suggestion
	globalThis.Prism?.constructor?.name === 'Object' ? globalThis.Prism.config : {};
```

DmitrySharabin commented on src/config.ts:
```diff
@@ -0,0 +1,65 @@
+	plugins: getGlobalArraySetting('plugins'),
+	languagePath: getGlobalSetting('language-path') ?? './languages/',
+	pluginPath: getGlobalSetting('plugin-path') ?? './plugins/',
+};
```
```suggestion
	languagePath: getGlobalSetting('language-path') as string | undefined ?? './languages/',
	pluginPath: getGlobalSetting('plugin-path') as string | undefined ?? './plugins/',
```

DmitrySharabin commented on src/config.ts:
```diff
@@ -0,0 +1,65 @@
+const hasDOM = typeof document !== 'undefined' && typeof window !== 'undefined';
+const scriptElement: HTMLScriptElement = hasDOM && document.currentScript;
+// @ts-ignore
+const globalConfig: Record<String, string | string[] | null> =
```
```suggestion
const globalConfig: Record<string, string | string[] | null> =
```

Ignore my suggestion if using `String` was intentional. If so, I'd love to know why. 🙂

DmitrySharabin commented on src/config.ts:
```diff
@@ -0,0 +1,65 @@
+const hasDOM = typeof document !== 'undefined' && typeof window !== 'undefined';
+const scriptElement: HTMLScriptElement = hasDOM && document.currentScript;
+// @ts-ignore
```
```suggestion
const scriptElement: HTMLOrSVGScriptElement | null = document?.currentScript;
```

From my perspective we get the same result with simpler types, no?

DmitrySharabin replied:
```suggestion
const scriptElement: HTMLOrSVGScriptElement | null = document?.currentScript;
```

From my perspective, we get the same result with a bit simpler types, no?

DmitrySharabin commented on src/config.ts:
```diff
@@ -0,0 +1,65 @@
+const hasDOM = typeof document !== 'undefined' && typeof window !== 'undefined';
+const scriptElement: HTMLScriptElement = hasDOM && document.currentScript;
```
```suggestion
const scriptElement: HTMLOrSVGScriptElement | null = document?.currentScript;
```

From my perspective, we get the same result with a bit simpler types, right?

DmitrySharabin commented on src/config.ts:
```diff
@@ -0,0 +1,65 @@
+	manual: getGlobalBooleanSetting('manual', !hasDOM),
+	languages: getGlobalArraySetting('languages'),
+	plugins: getGlobalArraySetting('plugins'),
+	languagePath: getGlobalSetting('language-path') ?? './languages/',
+	pluginPath: getGlobalSetting('plugin-path') ?? './plugins/',
```
```suggestion
	languagePath: getGlobalSetting('language-path') as string | null ?? './languages/',
	pluginPath: getGlobalSetting('plugin-path') as string | null ?? './plugins/',
```

DmitrySharabin commented on src/config.ts:
```diff
@@ -0,0 +1,65 @@
+	}
+	else if (hasDOM) {
+		return (
+			scriptElement.dataset[camelCaseName] ??
```
```suggestion
			scriptElement!.dataset[camelCaseName] ??
```

If I'm not mistaken, `scriptElement` will be defined here, but TS can't figure it out from the code, so we need to tell it we're sure with what we do.

DmitrySharabin commented on src/core/classes/prism.ts:
```diff
@@ -0,0 +1,132 @@
+import globalDefaults, { type PrismConfig } from './config';
```
```suggestion
import globalDefaults, { type PrismConfig } from '../../config';
```

DmitrySharabin commented on src/config.ts:
```diff
@@ -0,0 +1,65 @@
+	globalThis.Prism?.constructor?.name === 'Object' ? globalThis.Prism : {};
+
+function getGlobalSetting (name: string) {
+	let camelCaseName = name.replace(/-([a-z])/g, g => g[1].toUpperCase());
```
```suggestion
	let camelCaseName = name.replace(/-([a-z])/g, g => g[1].toUpperCase()); // eslint-disable-line regexp/no-unused-capturing-group
```

ESLint still can't analyze the code properly. 😅

DmitrySharabin commented on src/config.ts:
```diff
@@ -0,0 +1,65 @@
+}
+
+export const globalDefaults: PrismConfig = {
+	manual: getGlobalBooleanSetting('manual', !hasDOM),
```
```suggestion
	manual: getGlobalBooleanSetting('manual', !hasDOM),
	silent: getGlobalBooleanSetting('silent', false),
```

And here. `errorHandler` should probably stay `undefined`. It will be set in the Prism class constructor.

DmitrySharabin commented on src/core/classes/prism.ts:
```diff
@@ -0,0 +1,132 @@
+
+		const reportError: (reason: any) => PromiseLike<never> = this.config.errorHandler;
+
+		this.languageRegistry = new LanguageRegistry(this.config.languagePath ?? './languages/');
```
```suggestion
		this.languageRegistry = new LanguageRegistry(this.config.languagePath;
```

We probably don't need it since it's already in `globalDefaults`.

DmitrySharabin commented on src/core/classes/prism.ts:
```diff
@@ -0,0 +1,132 @@
+		const reportError: (reason: any) => PromiseLike<never> = this.config.errorHandler;
+
+		this.languageRegistry = new LanguageRegistry(this.config.languagePath ?? './languages/');
+		this.pluginRegistry = new PluginRegistry(this.config.pluginPath ?? './plugins/');
```
```suggestion
		this.pluginRegistry = new PluginRegistry(this.config.pluginPath);
```

The same is here.

DmitrySharabin commented on src/core/classes/prism.ts:
```diff
@@ -0,0 +1,132 @@
+
+		// Preload languages
+		if (this.config.languages?.length > 0) {
+			this.languageRegistry.loadAll(this.config.languages);
```
```suggestion
			this.languageRegistry.loadAll(this.config.languages!);
```

TS can't get that `languages` is not `undefined` here.

DmitrySharabin commented on src/core/classes/prism.ts:
```diff
@@ -0,0 +1,132 @@
+		if (this.config.plugins?.length > 0) {
+			let pluginsReady = this.languagesReady
+				.then(() => {
+					return this.pluginRegistry.loadAll(this.config.plugins);
```
```suggestion
					return this.pluginRegistry.loadAll(this.config.plugins!);
```

DmitrySharabin commented on src/core/classes/prism.ts:
```diff
@@ -0,0 +1,132 @@
+
+	constructor (config: PrismConfig = {}) {
+		this.config = Object.assign({}, globalDefaults, config);
+		this.config.errorHandler ??= this.config.silent ? () => undefined : console.error;
```
```suggestion
		this.config.errorHandler ??= (
			this.config.silent ? () => undefined : console.error
		) as PrismConfig['errorHandler'];
```

The code is already formatted with Prettier.

DmitrySharabin commented on src/core/classes/prism.ts:
```diff
@@ -0,0 +1,132 @@
+		this.config = Object.assign({}, globalDefaults, config);
+		this.config.errorHandler ??= this.config.silent ? () => undefined : console.error;
+
+		const reportError: (reason: any) => PromiseLike<never> = this.config.errorHandler;
```
```suggestion
		const reportError: PrismConfig['errorHandler'] = this.config.errorHandler;
```

DmitrySharabin commented on src/core/classes/prism.ts:
```diff
@@ -0,0 +1,132 @@
+		this.pluginRegistry = new PluginRegistry(this.config.pluginPath ?? './plugins/');
+
+		// Preload languages
+		if (this.config.languages?.length > 0) {
```
```suggestion
		// @ts-expect-error: TypeScript does not narrow the type of the languages property when using optional chaining in the condition.
		if (this.config.languages?.length > 0) {
```

TS drives me crazy sometimes. 🤪 We can either go the proposed way and ignore this error, since our code clearly shows our intention. Or we can help TS to narrow down types like so:

```ts
const languages = this.config.languages;
if (languages && languages.length > 0) {
	this.languageRegistry.loadAll(languages);
}
```

In that case, we can ignore the following suggestion (about adding `!`).

LeaVerou replied:
Let's do that (`languages && languages.length > 0`)

DmitrySharabin commented on src/config.ts:
```diff
@@ -0,0 +1,65 @@
+
+function getGlobalArraySetting (name: string): string[] {
+	const value = getGlobalSetting(name);
+	if (value === null || value === false || value === 'false') {
```
```suggestion
	if (value === undefined || value === false || value === 'false') {
```

DmitrySharabin commented on src/shared/hooks-util.ts:
```diff
@@ -1,4 +1,4 @@
-import type { HookCallback, HookEnvMap, Hooks } from '../core/hooks';
+import type { HookCallback, HookEnvMap, Hooks } from '../core/classes/hooks';
```
Is `HooksEnvMap` something to be added later? I can't find this type anywhere.

I can't add a suggestion to line 17, so show it here for not to lose:

```diff
- export function addHooks (hooks: Hooks, hooksMap: { [K in keyof HookEnvMap]?: HookCallback<K> }) {
+ export function addHooks (hooks: Hooks, hooksMap: { [K in keyof HookEnvMap]?: HookCallback }) {
```

LeaVerou replied:
I think it's part of the stuff that got removed.

DmitrySharabin commented on src/core/classes/hooks.ts:
```diff
@@ -0,0 +1,59 @@
+export class Hooks {
+	// eslint-disable-next-line func-call-spacing
```
```suggestion
```

No such error in the following line.

DmitrySharabin commented on src/types.d.ts:
```diff
@@ -18,10 +17,9 @@ export interface ComponentProtoBase<Id extends string = string> {
 export interface LanguageProto<Id extends string = string> extends ComponentProtoBase<Id> {
 	grammar: Grammar | ((options: GrammarOptions) => Grammar);
```
```suggestion
	grammar: Grammar | ((options?: GrammarOptions) => Grammar);
```

Otherwise, we can't do `grammar.call(this)` without passing mandatory parameters.

DmitrySharabin commented on src/core/classes/language.ts:
```diff
@@ -0,0 +1,78 @@
+				this.evaluatedGrammar = grammar.call(this);
+			}
+
+			if (this.evaluatedGrammar.$insertBefore) {
```
```suggestion
			if (this.evaluatedGrammar?.$insertBefore) {
```

DmitrySharabin commented on src/core/classes/language.ts:
```diff
@@ -0,0 +1,78 @@
+				delete this.evaluatedGrammar.$insertBefore;
+			}
+
+			if (this.evaluatedGrammar.$delete) {
```
```suggestion
			if (this.evaluatedGrammar?.$delete) {
```

DmitrySharabin commented on src/core/classes/language.ts:
```diff
@@ -0,0 +1,78 @@
+			// TODO extends should also be able to
+			let baseGrammar = def.extends ? { ...def.extends } : {};
+
+			if (typeof grammar === 'object') {
+				// the grammar is a simple object, so we don't need to evaluate it
+				this.evaluatedGrammar = Object.assign(baseGrammar, grammar);
+			}
+			else if (typeof grammar === 'function') {
+				this.evaluatedGrammar = grammar.call(this);
+			}
```
We use `else if` here, which means that `this.evaluatedGrammar` might be `undefined` after this block. It `undefined` is a possible value, we should fix the returned type to `Grammar | undefined`.

Do I get it right that if the `grammar` getter returns `undefined`, something is wrong?

DmitrySharabin commented on src/types.d.ts:
```diff
@@ -114,7 +112,11 @@ export interface GrammarSymbols {
-export type Grammar = GrammarTokens & GrammarSymbols;
+export interface GrammarSpecial {
+	$insertBefore?: GrammarTokens;
+	$delete: string[];
```
```suggestion
	$delete?: string[];
```

Otherwise, we can't do `delete grammar.$delete`.

DmitrySharabin commented on src/languages/c.ts:
```diff
@@ -5,9 +5,13 @@ import type { GrammarToken, LanguageProto } from '../types';
 	optional: 'opencl-extensions',
-	grammar ({ extend, getOptionalLanguage }) {
-		const c = extend('clike', {
+	grammar ({ getOptionalLanguage }) {
```
```suggestion
	grammar ({ getOptionalLanguage }: GrammarOptions) {
```

LeaVerou replied:
Can you do it?

DmitrySharabin commented on src/languages/c.ts:
```diff
@@ -28,78 +32,66 @@ export default {
+								pattern: /^(#\s*include\s*)<[^>]+>/,
+								lookbehind: true,
+							},
+							c['string'] as GrammarToken,
```
We should define `c` somehow. Previously, it was defined as `const c = extend(...)`.

LeaVerou replied:
Or eliminate the need to refer to it. 🤔

DmitrySharabin commented on src/languages/chaiscript.ts:
```diff
@@ -7,6 +7,7 @@ import type { LanguageProto } from '../types';
 	id: 'chaiscript',
 	require: [clike, cpp],
+	extends: clike,
 	grammar ({ extend, getLanguage }) {
```
```suggestion
	grammar ({ extend, getLanguage }: GrammarOptions): Grammar {
```

DmitrySharabin commented on src/plugins/toolbar/prism-toolbar.ts:
```diff
@@ -1,6 +1,6 @@
 import { getParentPre } from '../../shared/dom-util';
 import { noop } from '../../shared/util';
-import type { CompleteEnv, HookCallback } from '../../core/hooks';
+import type { CompleteEnv, HookCallback } from '../../core/classes/hooks';
```
Some new type not yet defined? I mean `CompleteEnv`.

LeaVerou replied:
No idea.

DmitrySharabin commented:
Code suggestion for `types.d.ts`, line 106:

```diff
- Record<TokenName, RegExpLike | GrammarToken | (RegExpLike | GrammarToken)[]>
+ Record<TokenName, RegExpLike | GrammarToken | GrammarTokens | (RegExpLike | GrammarToken | GrammarTokens)[]>
```

This allows us to cover cases like (in `languages/c.ts`, lines 35–42):

```ts
$insertBefore: {
				'string': {
					'char': {
						// https://en.cppreference.com/w/c/language/character_constant
						pattern: /'(?:\\(?:\r\n|[\s\S])|[^'\\\r\n]){0,32}'/,
						greedy: true,
					},
				},
...
```

DmitrySharabin commented on src/global.ts:
```diff
@@ -1,10 +1,8 @@
-// eslint-disable-next-line no-undef
-const namespace = globalThis as Partial<Record<typeof globalSymbol, Prism>>;
-const globalPrism = (namespace[globalSymbol] ??= new Prism());
+declare global {
+	const Prism: Prism | undefined;
+}
```
```suggestion
declare global {
	interface globalThis {
		Prism: Prism | undefined;
	}
}
```

This seems to be a better approach. However, I still don't understand why TS complains in `config.ts`. I tried adding the index signature to the `globalThis` interface, but it didn't help. 🤷‍♂️

```ts
declare global {
	interface globalThis {
		Prism: Prism | undefined;
		[key: string]: unknown;
	}
}
```

LeaVerou commented on src/types.d.ts:
```diff
@@ -18,10 +17,9 @@ export interface ComponentProtoBase<Id extends string = string> {
-type PluginType<Name extends string> = Name extends keyof KnownPlugins
-	? KnownPlugins[Name]
-	: unknown;
+type PluginType<Name extends string> = unknown;
```
Can you do it?

DmitrySharabin replied:
Sure. Just mark the changes you approve, and I’ll do that. 🙂

LeaVerou committed:
Fixes (#3915)

Update src/config.ts

Update src/core/classes/prism.ts

Update src/core/classes/hooks.ts

Update src/types.d.ts

Update src/languages/chaiscript.ts

Update src/global.ts

Update src/languages/java.ts

Update src/languages/c.ts

Update src/core/classes/language.ts

Update src/core/classes/language.ts

Update src/types.d.ts

Co-Authored-By: Dmitry Sharabin <dmitrysharabin@gmail.com>
Co-Authored-By: Lea Verou <lea@verou.me>

LeaVerou commented on src/config.ts:
```diff
@@ -0,0 +1,65 @@
+function getGlobalBooleanSetting (name: string, defaultValue: boolean): boolean {
+	const value = getGlobalSetting(name);
+
+	if (value === null) {
```
We should probably do both — a missing attribute will be `null`, not `undefined`.

LeaVerou commented on src/plugins/diff-highlight/prism-diff-highlight.ts:
```diff
@@ -1,7 +1,7 @@
 import diff, { PREFIXES } from '../../languages/diff';
 import { addHooks } from '../../shared/hooks-util';
-import type { BeforeSanityCheckEnv, BeforeTokenizeEnv } from '../../core/hooks';
+import type { BeforeSanityCheckEnv, BeforeTokenizeEnv } from '../../core/classes/hooks';
```
Leftovers I believe. Can you fix?

DmitrySharabin replied:
They are used to type the `env` parameter of a function. I think we need to define a type for `env` in one place and reference it in others, even if it is just `any`. Later, we can fix it once, and everything should just work. We have a lot of that stuff in plugins.

LeaVerou committed:
`extends` -> `base` and a few other things

LeaVerou committed:
Fix `c`

LeaVerou committed:
Concept for language extensions

LeaVerou committed:
Move $insertBefore and $delete to extend()

LeaVerou committed:
Separate language patches from other eval, attempt to move types to more relevant places

LeaVerou committed:
Update actionscript.ts

LeaVerou committed:
Update bison.ts

LeaVerou committed:
Type wrangling

LeaVerou committed:
Merge branch 'v2' into simplify

DmitrySharabin commented on src/core/classes/language.ts:
```diff
@@ -0,0 +1,96 @@
+		if (typeof grammar === 'function') {
+			grammar = grammar.call(this, {
+				base,
+				getLanguage (id: string) {
+					return this.registry.get(id);
+				},
```
I wonder if this should be an error function? Will `registry` be accessible if not?

LeaVerou replied:
Oh, yes it does need to be an arrow function! Feel free to commit directly to this PR

DmitrySharabin commented on src/core/classes/language.ts:
```diff
@@ -0,0 +1,96 @@
+	}
+
+	get resolvedGrammar () {
+		return (this.resolvedGrammar = resolveGrammar(this.grammar));
```
`resolveGrammar` mutates the passed grammar and doesn't return anything. Do we need to mutate `this.grammar`? If yes, then I can suggest we need something like:

```ts
		resolveGrammar(this.grammar);
		return (this.resolvedGrammar = this.grammar);
```

LeaVerou replied:
They were meant to be separate. Resolve grammar applies the combined patches (from `$insertBefore`, `$delete`, `$merge`) against the grammar. I'll need to think more about this

DmitrySharabin committed:
Some suggestions (#3919)

* Remove ESLint disable rule

* Better type compatibility

* Correctly re-export types

* Fix some types

DmitrySharabin committed:
Add proper typing for the `Grammar` type (#3918)

TS still treats it as `any` because of infinite recursion.

For now,  we limit the recursion depth to 5 (`GrammarTokensDepth`). We can change it if needed.

DmitrySharabin committed:
Preserve context

DmitrySharabin commented:
I tried to transform the ASP.NET (C#) language, and [here](#3920) is what I came up with. I used `self` to reference the grammar being defined. I know it's not a part of the new API, but I don't know how to reference the grammar inside the grammar definition the other way. I might miss something.

This was the trickiest part for me (and I'm still not sure if I did it correctly)

```ts
const tag = aspnet['tag'] as GrammarToken & {
	inside: { 'attr-value': { inside: Grammar } };
};

// Regexp copied from markup, with a negative look-ahead added
tag.pattern =
	/<(?!%)\/?[^\s>\/]+(?:\s+[^\s>\/=]+(?:=(?:("|')(?:\\[\s\S]|(?!\1)[^\\])*\1|[^\s'">=]+))?)*\s*\/?>/;

// match directives of attribute value foo="<% Bar %>"
insertBefore(tag.inside['attr-value'].inside, 'punctuation', {
	'directive': aspnet['directive'],
});
```

I transformed it into this:

```ts
$merge: {
	'tag': {
		// Regexp copied from markup, with a negative look-ahead added
		pattern: /<(?!%)\/?[^\s>\/]+(?:\s+[^\s>\/=]+(?:=(?:("|')(?:\\[\s\S]|(?!\1)[^\\])*\1|[^\s'">=]+))?)*\s*\/?>/,
		inside: {
			'attr-value': {
				inside: {
					$insertBefore: {
						'punctuation': {
							// match directives of attribute value foo="<% Bar %>"
							'directive': self['directive'],
						},
					}
				}
			}
		}
	}
},
```

DmitrySharabin commented on src/core/classes/language.ts:
```diff
@@ -0,0 +1,96 @@
+
+		if (typeof grammar === 'function') {
+			grammar = grammar.call(this, {
+				base,
```
We should probably pass the required languages as well, so we can destructure them in the `grammar()` function and use the way we can use `base`. It can be helpful, for example, when defining the `JSDoc` language: there will be no need to get the needed languages using `getLanguage()`. See https://github.com/PrismJS/prism/blob/44429cac5aa9e2ae56d8adab20195edb1496fd30/src/languages/jsdoc.ts#L11-12.

LeaVerou committed:
Merge branch 'simplify' of https://github.com/PrismJS/prism into simplify

LeaVerou committed:
Move iterable helpers to separate file

LeaVerou committed:
TS

LeaVerou committed:
Merge branch 'v2' into simplify

LeaVerou committed:
Update eslint.config.mjs

LeaVerou committed:
WIP, improve API around languages

DmitrySharabin committed:
First pass on transforming `php` and `php-extras` (#3922)

Co-authored-by: Lea Verou <lea@verou.me>

DmitrySharabin committed:
First pass on transforming ASP.NET (C#) (#3920)

DmitrySharabin committed:
First pass (#3921)

LeaVerou committed:
Move grammar back into types

LeaVerou committed:
Languages

LeaVerou committed:
Drop tokenize and rest symbols in favor of $tokenize and $rest properties

Also simplify lang helpers

LeaVerou committed:
Update tokenize.ts

LeaVerou committed:
Sample rewrite to show what group matching + $language can do

LeaVerou committed:
Remove effect from languages

DmitrySharabin commented on src/shared/language-util.ts:
```diff
@@ -104,100 +105,82 @@ export function insertBefore (grammar: Grammar, before: string, insert: GrammarT
  */
-export function extend (grammar: Grammar, id: string, reDef: Grammar): Grammar {
-	const lang = cloneGrammar(grammar, id);
+export function extend (base: Grammar, id: string, grammar: Grammar): Grammar {
```
We can probably ditch `id` here (it's not used anywhere in the function body). `deepClone()` doesn't need it as a parameter anymore.

DmitrySharabin committed:
Suggestions to the `simplify` branch (#3925)

LeaVerou committed:
Allow implicit any

The whole point of TS is that it works alongside regular JS. noImplicitAny defeats this benefit.

LeaVerou committed:
Move language utils to `util/` and separate them out

LeaVerou committed:
Add aliases to C++, C#, F# so markdown doesn't need to handle them specially

LeaVerou committed:
Drop xml-doc

LeaVerou committed:
Drop shared file, not worth it

LeaVerou committed:
Some more work on dynamic langs (still very WIP)

LeaVerou committed:
`resolveGrammar` -> `grammarPatch`

LeaVerou committed:
insertAfter, deep keys

LeaVerou committed:
Make php-extras just a part of PHP

LeaVerou committed:
Add media types and extensions

DmitrySharabin committed:
Fix typo

LeaVerou commented on src/core/classes/language.ts:
```diff
@@ -0,0 +1,177 @@
+		}
+
+		for (let def of this.require) {
+			Object.defineProperty(this.languages, def.id, {
+				enumerable: true,
+				configurable: true,
+				get: () => {
+					let language = this.registry.peek(def as LanguageProto);
+					if (language) {
+						// Already resolved
+						this.languages[def.id] = language;
+					}
+					else {
+						this.registry.add(def as LanguageProto);
+						defineLazyProperty(this.languages, def.id, () => {
+							return this.registry.getLanguage(def.id)!;
+						});
+					}
+				},
+			});
```
Oh that's clearly a mistake, yes. I think the outer defineProperty() can just be removed.

DmitrySharabin replied:
Ok, I'll submit a PR so we are good with that change.

DmitrySharabin commented on src/core/highlight.ts:
```diff
@@ -0,0 +1,55 @@
+	language: string,
+	options: HighlightOptions = {}
+): string {
+	const prism = this ?? singleton;
```
We introduce a cycle by importing `singleton` in all' highlight' methods. This is how it looks:

```
src/core/prism.ts
  └──> src/core/classes/prism.ts
         └──> src/core/highlight.ts
                └──> src/core/prism.ts
```

Moving it to the file with the Prism class won't help. We probably need to break it somehow. Possibly, this is not only the issue in the build step. It might cause some other issues with types? 🤔

LeaVerou replied:
Cycles are not a problem in ESM, they happen all the time. Does Rollup still struggle with them? I seem to recall it had an issue in the past.

DmitrySharabin replied:
It logs a warning, but builds the code fine.

LeaVerou replied:
Cool, no issue then? :D

DmitrySharabin committed:
Some more suggestions to the `simplify` branch (#3926)

Co-authored-by: Lea Verou <lea@verou.me>

LeaVerou committed:
Preserve accessors when extending/copying objects

That way we can have more flexibility with lazy properties etc

DmitrySharabin committed:
Suggestions (next iteration) (#3928)

LeaVerou committed:
Simplify hooks, add type safety

LeaVerou committed:
Update util.ts

LeaVerou committed:
WIP

- Make `resolvedGrammar` always exist
- Rewrite `highlight` to take advantage of LanguageRegistry
- Define hook types
- Move token types to Token

DmitrySharabin commented on src/core/classes/language-registry.ts:
```diff
@@ -0,0 +1,114 @@
+
+		id = this.aliases[id] ?? id;
+		def ??= this.cache[id];
+
+		return { id, def };
```
```suggestion
		let language = this.instances[id];

		return { id, def, language };
```

If there's already a resolved language corresponding to `id`, we should probably return it as well.

DmitrySharabin commented on src/core/highlight.ts:
```diff
@@ -0,0 +1,70 @@
+
+	const { id, def, language } = prism.languageRegistry.resolveRef(languageRef);
+
+	let env: HookEnv['after-tokenize'] = {
```
I think it's a typo: we define `env` right before we run the `before-tokenize` hook.

DmitrySharabin replied:
Oh, I see. So, it just should be a variable that covers all cases. Got it. My suggestion is irrelevant then.

DmitrySharabin commented on src/core/highlight.ts:
```diff
@@ -0,0 +1,70 @@
+
+	prism.hooks.run('before-tokenize', env);
+
+	if (!env.language) {
+		throw new Error(`No language definition found for ${env.languageId}.`);
+	}
```
Does this mean that `language` can be added to `env` during the `before-tokenize` hook run? Otherwise, we probably would have performed this check before it.

LeaVerou replied:
Yes

DmitrySharabin commented on src/core/highlight.ts:
```diff
@@ -0,0 +1,70 @@
+	}
+
+	env = env as unknown as HookEnv['after-tokenize'];
+	env.tokens = tokenize.call(prism, env.code, env.language!.resolvedGrammar);
```
It's only for TS, right? As an alternative, I'd suggest explicitly adding a `token` property to `env`:

DmitrySharabin replied:
It's only for TS, right? As an alternative, I'd suggest explicitly adding a `token` property to `env`:

```suggestion
	env = { ...env, tokens: undefined } as HookEnv['after-tokenize'];
```

DmitrySharabin commented on src/core/highlight.ts:
```diff
@@ -24,32 +40,31 @@ import type { Grammar, GrammarToken, GrammarTokens, RegExpLike } from '../types'
 	}
 
-	env.tokens = prism.tokenize(env.code, env.grammar.resolvedGrammar);
+	env = env as unknown as HookEnv['after-tokenize'];
```
It's only for TS, right? As an alternative, I'd suggest explicitly adding a `token` property to `env`:

```suggestion
	env = { ...env, tokens: undefined } as HookEnv['after-tokenize'];
```

DmitrySharabin commented on src/core/highlight.ts:
```diff
@@ -24,32 +40,31 @@ import type { Grammar, GrammarToken, GrammarTokens, RegExpLike } from '../types'
 
-	env.tokens = prism.tokenize(env.code, env.grammar.resolvedGrammar);
+	env = env as unknown as HookEnv['after-tokenize'];
+	env.tokens = tokenize.call(prism, env.code, env.language!.resolvedGrammar);
```
Do we really need to use `call` here? `tokenize` already expects a Prism instance as the first argument. Or do you have some changes to its signature in mind?

DmitrySharabin replied:
Or probably it's a TIL thing: if we call the first argument of a function `this`, we can use it inside the function body to assign some default context the function will work with? And to pass this context explicitly, we use `call`?

LeaVerou replied:
It's definitely confusing syntax though!

LeaVerou commented on src/languages/c.ts:
```diff
@@ -1,22 +1,31 @@
 
 export default {
 	id: 'c',
 	require: clike,
```
Correct.

DmitrySharabin commented on src/core/classes/hooks.ts:
```diff
@@ -0,0 +1,104 @@
+			let hooks = name;
+
+			for (let name in hooks) {
+				this.add(name, hooks[name]);
```
```suggestion
				this.add(name, hooks[name as string]);
```

LeaVerou committed:
Update src/core/classes/language-registry.ts

Co-authored-by: Dmitry Sharabin <dmitrysharabin@gmail.com>

LeaVerou committed:
Update highlight.ts

LeaVerou committed:
Merge branch 'simplify' of https://github.com/PrismJS/prism into simplify

LeaVerou committed:
Update src/core/classes/hooks.ts

Co-authored-by: Dmitry Sharabin <dmitrysharabin@gmail.com>

LeaVerou committed:
Drop `$language`

LeaVerou committed:
Break down `tokenize.ts` into separate modules

LeaVerou committed:
Prettier

LeaVerou committed:
Registry -> ComponentRegistry

LeaVerou committed:
async util

DmitrySharabin commented on src/languages/coffeescript.ts:
```diff
@@ -38,67 +37,58 @@ export default {
+								pattern: /^`|`$/,
+								alias: 'punctuation',
+							},
+							'script': 'javascript',
```
Shouldn't that be something like

```js
'script': { $inside: 'javascript` }
```

LeaVerou replied:
No, I think that happens automatically. Or at least it should (though I’m starting to think special casing languages as strings is not a good idea but let’s get the branch merged first before changing that)

DmitrySharabin replied:
Okay. But what is the difference between `inside` and `$inside`? Is `inside` something that should be replaced with `$inside` eventually (like it happened with `rest` that became `$rest`)?

LeaVerou replied:
Yes, we should add the `language-xxx` class automatically.

LeaVerou committed:
Update eslint.config.mjs

LeaVerou committed:
Update async.ts

LeaVerou committed:
Update async.ts

LeaVerou committed:
Update language.ts

LeaVerou committed:
Update language.ts

LeaVerou committed:
Update types.d.ts

LeaVerou committed:
tokenize

LeaVerou committed:
Move Token class to classes

LeaVerou committed:
First stab at functional $inside/$rest

cc @DmitrySharabin

DmitrySharabin commented on src/core/tokenize/match.ts:
```diff
@@ -46,11 +42,21 @@ export function _matchGrammar (
-				patternObj.pattern = pattern = RegExp(pattern.source, pattern.flags + 'g');
+				flagsToAdd += 'g';
+			}
+			if (pattern.source.indexOf('(?<') && pattern.hasIndices === false) {
```
Looking at the expression, it seems that `includes` is intended to be used instead of `indexOf`, no?

LeaVerou replied:
Yes

DmitrySharabin commented on src/core/tokenize/match.ts:
```diff
@@ -146,13 +153,37 @@ export function _matchGrammar (
+							content = tokenize.call(prism, content, localInsideGrammar);
+						}
+
+						return typeof arg === 'object' && arg.type
```
Why not `type` instead of `arg.type` here and in the following line?

LeaVerou replied:
TS was complaining

DmitrySharabin committed:
Some suggestions to the `simplify` branch (#3939)

* Types and type safety

* Fix the build script

* Remove unnecessary import

* Import lost types + Prettier

* Use `includes` instead of `indexOf`

* Add type for better auto completion

DmitrySharabin commented on src/core/classes/component-registry.ts:
```diff
@@ -0,0 +1,171 @@
+	 */
+	private loadingList: Promise<T>[] = [];
+
+	ready: Promise<T[]>;
```
Shouldn't `ready` be a getter to become reactive? We change dynamically `loadingList` (`ready` depends on it). However, we set the value of `ready` once when creating an instance of a registry and never update it. As a result, we have `ready` that is always fulfilled, even though there might be promises that are not fulfilled in the `loadingList` array.
Or should we update `ready` every time we update `loadingList`? Or is there another option I'm not seeing?

DmitrySharabin committed:
Update the `Language` class (#3941)

- If a grammar has a base language, it should always be resolved
- Fix `getLanguage()`: `language` can already be a resolved grammar if it is taken from `this.languages`

DmitrySharabin committed:
Some fixes to `tokenize` (#3942)

- Don’t choke on patterns that are not instances of `RegExp`
- In `resolve()` return `resolvedGrammar` and delete `$rest` after it is properly resolved

DmitrySharabin committed:
Partially fix tests (in the `simplify` branch) (#3940)

* Fix `test:components`

* Partially adopt the new API

* Fix `test:identifiers`

* Use `languageRegistry`

* Use cache

* Fix core tests

* Fix ESLint and TS errors + typo in `regex-coverage`

* Fix ESLint and TS errors in `test:patterns`

DmitrySharabin commented:
Things to implement before merging:

- [ ] `$self`
- [ ] Correctly resolve `inside: 'language-id'`

Something else?

DmitrySharabin committed:
Another attempt to avoid the “too deep” structure  error and not fall back to `any` (#3944)

LeaVerou commented:
For `$self`, we need to figure out what that means when we have nested languages (e.g. highlighting JS in HTML). Probably the closest language.

`inside: 'language-id'` worked before, so that must be a regression.

DmitrySharabin commented:
> `inside: 'language-id'` worked before, so that must be a regression.

Yup, this is related to something we've already [started to discuss](https://github.com/PrismJS/prism/pull/3942#discussion_r2106217545).

DmitrySharabin committed:
Update Prettier plugins (#3945)

We had an [issue](https://github.com/ony3000/prettier-plugin-merge/issues/36) formatting complex files with many regexes, which is now fixed.

LeaVerou commented:
> > `inside: 'language-id'` worked before, so that must be a regression.
> 
> Yup, this is related to something we've already [started to discuss](https://github.com/PrismJS/prism/pull/3942#discussion_r2106217545).

That does not address why it stopped working. This was even working in v1. And the code to make it work is still there I believe (in tokenize)

DmitrySharabin committed:
Bug fixes (the `simplify` branch) (#3943)

* [C] Fix transformation bug: `macro` should be inserted before `string`

* [F#] Specify the base language

* [extend] Deeply nested objects can be `undefined`

* [C#] Specify the base language + remove unused imports

* [PHP] Add missed tokens

* [VB.NET] Fix language definition

* [MD] Correctly capture the language code

DmitrySharabin committed:
Correctly parse CLI arguments (#3949)

DmitrySharabin committed:
Make `nextTick()` work in environments other than browsers (#3952)

DmitrySharabin committed:
Resolve the language if its definition is already loaded, and it is not yet resolved (#3951)

DmitrySharabin committed:
Fix the Markdown language (#3950)

- Add the missing inside token and fix the existing one
- Fix the regex that is too greedy now: if we have more than one code block next to each other, all of them starting from the second one will be interpreted as a code block of the first one
- Ignore spaces before the language code

DmitrySharabin committed:
Transform languages (#3948)

* Vala

* Firestore security rules

* Xeora

* Flow

* Wiki

* V

* Velocity

* glsl

* gml

* Go

* gradle

* groovy

* haxe

* kotlin

* javadoc

* squirrel

* hlsl

* idris

* json5

* jsonp

* less

* n4js

* objectivec

* opencl

* sqf

* sparql

* solidity

* scala

* reason

* racket

* qore

* purescript

* purebasic

* protobuf

* processing

* plsql

* phpdoc

* mongodb

* jolie

* latte

* parser

* textile

* qsharp

* ruby

* sass

* scss

* xquery

* http

* haml

* rescript

DmitrySharabin committed:
Fix the gradle language (#3955)

DmitrySharabin committed:
In addition to waiting for all promises to settle, handle post-hoc additions/removals (#3961)

Credit: https://github.com/mavoweb/mavo/blob/78efe2b9cadd09c1d131b8afd5fe2f38d5cfa8c7/src/mavo.js#L845

Partially addresses #3960

DmitrySharabin committed:
Integrate changes from the `v2` branch (#3982)

* Place plugin docs along with their source code (#3930)

* [autoloader] Fix script name in the docs

* [normalize-whitespace] Don't consider demo as a plugin

* Generate `file-sizes.json` during the build (#3946)

* Exclude plugins' `demo.js` files from type checking

* Fix path to the core entry point in `components.json`

* Move and rename themes (#3947)

- Move themes to the `src` folder and remove the `prism-` prefix
- Adjust the build process accordingly
- Adjust themes' metadata:
    - Remove broken links
    - Add/update authors' full names
    - Update GitHub handles
    - Use Markdown-style links
    - In `components.json`, fix theme ids and add Lea Verou as the owner to the prism, dark, and funky themes

* Update themes' docs (#3958)

So that they follow the structure (we can parse and use on the website):

```
/**
 * <title>
 *
 * <description>
 *
 * @author <name> (<GitHub_username>)
 */
```

Between the data chunks, there should be at least one empty line.

* Move `combineCallbacks` to its own module (#3963)

* Move the `Token` class under the `classes` folder

* Move all utils for working with iterables to a separate module

Remove unused utils

* Simplify hooks

* Move the `Prism` class, add Prism singleton

* Simplify the `Prism` class

Move to separate modules:
- `highlightAll()`
- `highlightElement()`
- `highlight()`
- `stringify()`

* Correctly parse CLI arguments (#3972)

Was fixed in #3949 (in the `simplify` branch). Move it here to make testing easier.

* Simplify the `Prism` class even more

- Move `tokenize()`, `_matchGrammar()` (and related code), and `resolve()` to separate modules.
- Ditch the `tokenize` symbol in favor of the `$tokenize` special property. Since symbols are excluded from the `for…in` loop, and `$tokenize` is not a symbol, to preserve the previous behavior, we need to handle it in `_matchGrammar` explicitly.
- Switch from `[tokenize]` to `$tokenize` in languages.

---
Co-authored-by: Dmitry Sharabin <dmitrysharabin@gmail.com>

* Ditch the `rest` symbol in favor of the `$rest` special property

* Move language utils to `util/` and separate them out

* Remove known plugins + type safety

* Move `documentReady` to async utils

* Make the global bundle of Prism v2 work the same way as Prism v1 does

Now, the user can include Prism the way they used to (except v2 now uses the Autoloader plugin under the hood, so more languages can be highlighted right out of the box):

```html
<link rel="stylesheet" href="themes/prism.css" />
<script src="prism.js"></script>
```

Besides, they have access to the Prism instance via `globalThis`.

* Move Prism config to a separate file and fix the build options of the global bundle

Co-Authored-By: Dmitry Sharabin <dmitrysharabin@gmail.com>

* Fix some code after merging

---------

Co-authored-by: Lea Verou <lea@verou.me>

DmitrySharabin committed:
Merge branch 'v2' into simplify

DmitrySharabin committed:
Fix issues after resolving conflicts
